Replace more use of MallocPtr with newer MallocSpan

cdumez opened the pull request:
#### 469e8d288d11f72a19c28830fd431ef07fcb0567
<pre>
Replace more use of MallocPtr with newer MallocSpan
<a href="https://bugs.webkit.org/show_bug.cgi?id=282392">https://bugs.webkit.org/show_bug.cgi?id=282392</a>

Reviewed by Geoffrey Garen and Darin Adler.

Replace more use of MallocPtr with newer MallocSpan.

Also fix an issue when using MallocSpan&lt;T&gt; when sizeof(T)
is not 1. We were failing to multiply the size by sizeof(T)
before calling malloc functions.

* Source/JavaScriptCore/bytecode/ExpressionInfo.h:
* Source/JavaScriptCore/jit/JITCodeMap.h:
* Source/JavaScriptCore/runtime/VM.h:
* Source/WTF/wtf/Vector.h:
(WTF::VectorBufferBase::releaseBuffer):
(WTF::VectorBuffer::releaseBuffer):
(WTF::Malloc&gt;::releaseBuffer):
* Source/WTF/wtf/cocoa/VectorCocoa.h:
(WTF::dispatchDataFromVector):
* Source/WTF/wtf/text/StringBuffer.h:
(WTF::StringBuffer::release):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::StringImpl):
* Source/WebCore/platform/network/cf/FormDataStreamCFNet.cpp:
(WebCore::closeCurrentStream):
(WebCore::advanceCurrentStream):
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.mm:
(WebKit::NetworkRTCTCPSocketCocoa::sendTo):
(WebKit::dataFromVector): Deleted.
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSendStreamCocoa.mm:
(WebKit::NetworkTransportSendStream::sendBytes):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::sendDatagram):
* Source/WebKit/Shared/API/APIData.h:
(API::Data::createWithoutCopying):
(API::Data::create):
(API::Data::Data):
* Source/WebKit/Shared/Daemon/DaemonUtilities.mm:
(WebKit::vectorToXPCData):
* Tools/TestWebKitAPI/NetworkConnection.h:
* Tools/TestWebKitAPI/NetworkConnection.mm:
(TestWebKitAPI::Connection::awaitableSend):
(TestWebKitAPI::Connection::send const):
(TestWebKitAPI::Connection::sendAndReportError const):
(TestWebKitAPI::dataFromVector): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAlternateHTMLString.mm:
(TEST(WebKit, LoadMoreThan4GB)):

Canonical link: <a href="https://commits.webkit.org/286028@main">https://commits.webkit.org/286028@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/53588a4ce36039f69fde33b13257b3ad0f04cdc6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74545 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53974 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27356 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/78983 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25783 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76662 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63107 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1759 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/58596 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16878 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77612 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48737 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64095 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38984 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45806 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21591 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24116 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/67682 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67147 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21937 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80454 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/73803 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1862 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1099 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66858 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2010 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64113 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66145 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10078 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/8236 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/95584 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11505 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1827 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/4614 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20983 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1855 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/1878 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1862 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WTF/wtf/MallocSpan.h:
```diff
@@ -67,33 +67,33 @@ template<typename T, typename Malloc = FastMalloc> class MallocSpan {
     static MallocSpan malloc(size_t size)
     {
-        return MallocSpan { static_cast<T*>(Malloc::malloc(size)), size };
+        return MallocSpan { static_cast<T*>(Malloc::malloc(size * sizeof(T))), size };
```
Probably it would be a good idea to use `wtf/CheckedArithmetic.h` to ensure that the multiplication doesn't overflow, something like:

```suggestion
        Checked<size_t> allocationSize = sizeof(T);
        allocationSize *= size; 
        return MallocSpan { static_cast<T*>(Malloc::malloc(allocationSize), size };
```

aperezdc commented on Source/WTF/wtf/MallocSpan.h:
```diff
@@ -67,33 +67,33 @@ template<typename T, typename Malloc = FastMalloc> class MallocSpan {
     static MallocSpan tryMalloc(size_t size)
     {
-        auto* ptr = Malloc::tryMalloc(size);
+        auto* ptr = Malloc::tryMalloc(size * sizeof(T));
```
Same as above, but IIUC in this case we would want `CheckdSize` instead of plain `Checked<size_t>` and using its `.hasOverflowed()` method:

```suggestion
        CheckedSize allocationSize = sizeof(T);
        allocationSize *= size;
        if (allocationSize.hasOverflowed())
            return nullptr;
        auto* ptr = Malloc::tryMalloc(allocationSize);
```

aperezdc reviewed (changes requested):
I'm not very familiar with the Cocoa bits, but overall the patch looks good. Also I think it would be good to check for multiplication overflows before allocating memory, I left a couple of notes about that.

aperezdc commented on Source/WTF/wtf/MallocSpan.h:
```diff
@@ -67,33 +68,39 @@ template<typename T, typename Malloc = FastMalloc> class MallocSpan {
     static MallocSpan malloc(size_t size)
     {
-        return MallocSpan { static_cast<T*>(Malloc::malloc(size)), size };
+        return MallocSpan { static_cast<T*>(Malloc::malloc(CheckedSize(size) * sizeof(T))), size };
```
Well, there's an alias `using CheckedSize = Checked<size_t, RecordOverflow>`, which detects overflow _without_ aborting on overflow (because of the `RecordOverflow`). That's why I suggested `CheckedSize` for the `try[Zeroed]{M,Re}alloc` functions.

Here, for the non-`try*` functions, we should be using `Checked<size_t>`, which is effectively `Checked<size_t, CrashOnOverflow>` because `CrashOnOverflow` is the default template argument value.

cdumez replied:
I still think my code is correct? My code will crash as expected thanks to this logic here:
```
    operator T() const
    {
        if (UNLIKELY(this->hasOverflowed()))
            this->crash();
        return m_value;
    }
```

My CheckedSize gets implicitly converted to a size_t to pass to malloc(), and because it has overflowed, the conversion operator will call crash().

darinadler replied:
Yes, the design of `Checked` is strange, and it’s actually almost never important to use `CrashOnOverflow`, despite how the name sounds. We still end up crashing if there is overflow even without that flag. Might be nice to simplify `Checked` some day to avoid that confusion.

darinadler replied:
I think it’s subtle to change the semantic here without changing names. It’s non-traditional for a function named `malloc` to take a count of objects rather than a count of bytes. Many C++ programmers know that `malloc` takes a byte count and `calloc` and `new` take an object count. I can imagine callers passing a byte count, and then we’d make a too-large allocation, and we might never even notice since it’s wasteful but otherwise harmless.

I suggest we consider changing these function names to not literally contain `malloc` when making this change.

cdumez replied:
Do you prefer function renames to keeping `malloc()` and taking in a `sizeInBytes` instead?

darinadler replied:
I don’t have a strong preference either way.

aperezdc replied:
@cdumez Ah, you're right, I managed to miss the `this->crash()` in the conversion operator. I agree with Darin that the `Checked` design is indeed a bit odd, it fooled me a bit today after all!

geoffreygaren replied:
FWIW, I prefer sizeInBytes (as you've done it in your most recent patch). For clients who actually want "malloc" for whatever reason, sizeInBytes is how "malloc" works. For clients who want objects of type T, probably they should use FixedVector, which takes a count in number of objects and also performs initialization and destruction.

cdumez replied:
I like the latest iteration a lot more too. I think it is less error prone.

aperezdc reviewed (changes requested):
There's a small wrinkle, the functions without the `try*` prefix should use `Checked<size_t>` directly, see the comment below.

I like the `CheckedSize(size) * sizeof(T)` shorthand you used, it makes the code a bit more readable 👌🏼

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -1805,21 +1802,20 @@ void Vector<T, inlineCapacity, OverflowHandler, minCapacity, Malloc>::appendCont
-        size_t bytes = m_size * sizeof(T);
-        buffer = adoptMallocPtr<T, Malloc>(static_cast<T*>(Malloc::malloc(bytes)));
-        memcpy(buffer.get(), data(), bytes);
+        buffer = MallocSpan<T, Malloc>::malloc(m_size * sizeof(T));
```
I think this is the very mistake I was worried about. We need to remove the `* sizeof(T)` here, right?

cdumez replied:
Sure looks like it :(

darinadler commented on Source/WTF/wtf/cocoa/VectorCocoa.h:
```diff
@@ -118,7 +119,17 @@ inline Vector<uint8_t> makeVector(NSData *data)
+    auto vectorSize = vector.size();
+    auto vectorBuffer = vector.releaseBuffer();
+    auto* bufferPtr = vectorBuffer.span().data();
+    return adoptNS(dispatch_data_create(bufferPtr, vectorSize * sizeof(T), dispatch_get_main_queue(), makeBlockPtr([vectorBuffer = WTFMove(vectorBuffer)] { }).get()));
```
We know that this multiplication won’t overflow, but it looks like it could, and it would be nice to have a style where it doesn’t look that way. Maybe we should use `byte_size()` instead? Can make the local variable be `bufferSpan` or `buffer` and write `buffer.data(), buffer.byte_size()`.

cdumez replied:
Ok, going with:
```
template<typename T>
inline RetainPtr<dispatch_data_t> dispatchDataFromVector(Vector<T>&& vector)
{
    auto vectorSize = vector.size();
    auto vectorBuffer = vector.releaseBuffer();
    auto bufferSpan = vectorBuffer.span().first(vectorSize);
    return adoptNS(dispatch_data_create(bufferSpan.data(), bufferSpan.size_bytes(), dispatch_get_main_queue(), makeBlockPtr([vectorBuffer = WTFMove(vectorBuffer)] { }).get()));
}
```

darinadler replied:
Does this mean that `releaseBuffer` creates a span with a size matching the vector _capacity_, not the vector _size_? I’m not sure we need to do it that way. If the span was the vector size, then we wouldn’t need the call to `first`.

cdumez replied:
Yes, this is the behavior I have implemented in this PR. It is the true buffer size, which is not necessarily the same as the size of the vector (due to extra capacity). I agree though that this is a little error-prone so maybe we should indeed have have the MallocSpan size be the size of the vector.

darinadler replied:
I think we ought to start by using the size of the vector because we don’t typically make changes to the size after releasing the buffer. If we run into any case where we want to release the buffer and then later take that additional capacity into account, we can easily make a second release function for that.

darinadler commented on Source/WTF/wtf/cocoa/VectorCocoa.h:
```diff
@@ -118,7 +119,17 @@ inline Vector<uint8_t> makeVector(NSData *data)
 }
 
+template<typename T>
+inline RetainPtr<dispatch_data_t> dispatchDataFromVector(Vector<T>&& vector)
```
Not sure that I’m a fan of having the words _from vector_ in the function name. I would call this `makeDispatchData`. Doesn’t seem likely that it would compile if it was misused by a caller that’s not moving a `Vector`. I like to use the word _make_ rather than _create_ for functions that safely create something without potentially leaking because they use a smart pointer or other object that manages storage.

darinadler replied:
Not sure about my _make_ vs. _create_ comment, on reflection. Either might be good.

darinadler commented on Source/WebCore/platform/network/cf/FormDataStreamCFNet.cpp:
```diff
@@ -132,8 +132,8 @@ static bool advanceCurrentStream(FormStreamFields* form)
-            MallocPtr<uint8_t, WTF::VectorBufferMalloc> data = bytes.releaseBuffer();
-            form->currentStream = adoptCF(CFReadStreamCreateWithBytesNoCopy(0, data.get(), size, kCFAllocatorNull));
+            auto data = bytes.releaseBuffer();
+            form->currentStream = adoptCF(CFReadStreamCreateWithBytesNoCopy(0, data.span().data(), size, kCFAllocatorNull));
```
I’d be tempted to just write out `bytes.releaseBuffer().span().data()` without a local variable.

cdumez replied:
The local variable is used on the next line though.

darinadler replied:
Oops, my mistake!

darinadler commented on Source/WebKit/Shared/Daemon/DaemonUtilities.mm:
```diff
@@ -79,11 +80,7 @@ void startListeningForMachServiceConnections(const char* serviceName, ASCIILiter
-        fastFree(rawPointer);
-    }));
+    RetainPtr dispatchData = dispatchDataFromVector(WTFMove(vector));
     return adoptNS(xpc_data_create_with_dispatch_data(dispatchData.get()));
```
Would be nicer without the local variable:
```
return adoptNS(xpc_data_create_with_dispatch_data(makeDispatchData(WTFMove(vector)).get()));
```

darinadler reviewed (changes requested):
Love these changes! I found one size calculation mistake, though.

geoffreygaren reviewed (approved):
r=me

cdumez committed:
Replace more use of MallocPtr with newer MallocSpan
https://bugs.webkit.org/show_bug.cgi?id=282392

Reviewed by Geoffrey Garen and Darin Adler.

Replace more use of MallocPtr with newer MallocSpan.

Also fix an issue when using MallocSpan<T> when sizeof(T)
is not 1. We were failing to multiply the size by sizeof(T)
before calling malloc functions.

* Source/JavaScriptCore/bytecode/ExpressionInfo.h:
* Source/JavaScriptCore/jit/JITCodeMap.h:
* Source/JavaScriptCore/runtime/VM.h:
* Source/WTF/wtf/Vector.h:
(WTF::VectorBufferBase::releaseBuffer):
(WTF::VectorBuffer::releaseBuffer):
(WTF::Malloc>::releaseBuffer):
* Source/WTF/wtf/cocoa/VectorCocoa.h:
(WTF::dispatchDataFromVector):
* Source/WTF/wtf/text/StringBuffer.h:
(WTF::StringBuffer::release):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::StringImpl):
* Source/WebCore/platform/network/cf/FormDataStreamCFNet.cpp:
(WebCore::closeCurrentStream):
(WebCore::advanceCurrentStream):
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.mm:
(WebKit::NetworkRTCTCPSocketCocoa::sendTo):
(WebKit::dataFromVector): Deleted.
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSendStreamCocoa.mm:
(WebKit::NetworkTransportSendStream::sendBytes):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::sendDatagram):
* Source/WebKit/Shared/API/APIData.h:
(API::Data::createWithoutCopying):
(API::Data::create):
(API::Data::Data):
* Source/WebKit/Shared/Daemon/DaemonUtilities.mm:
(WebKit::vectorToXPCData):
* Tools/TestWebKitAPI/NetworkConnection.h:
* Tools/TestWebKitAPI/NetworkConnection.mm:
(TestWebKitAPI::Connection::awaitableSend):
(TestWebKitAPI::Connection::send const):
(TestWebKitAPI::Connection::sendAndReportError const):
(TestWebKitAPI::dataFromVector): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAlternateHTMLString.mm:
(TEST(WebKit, LoadMoreThan4GB)):

Canonical link: https://commits.webkit.org/286028@main
